### ISSUE TITLE
Remove DBs when the schema of the downloaded DB is older than the current schema

### ISSUE THREAD

**Issue by marco-c** [mozilla/bugbug#1108]
In the download method, https://github.com/mozilla/bugbug/blob/d1016a351603d948230188fa3f370a057a1a8127/bugbug/db.py#L89, we should handle the case where the DB is already downloaded but is using an older schema version than the current one.

**Comment by marco-c** [comment]
It's possible that this is already kind of supported, as we check the ETag and the ETag changes when the schema version changes. We need to add a test to confirm.

**Comment by Nathan-Wisner** [comment]
For this, would we want to update and download a newer version or just exit the function with false like 
``` if is_old_schema(path):
     return False
```

**Comment by marco-c** [comment]
We should try to download a new version.

**Comment by mohanish2504** [comment]
Hi,
I want to contribute and I am a beginner, I wanted to know is the issue resolved? I am stuck here 

'if is_old_schema(path): return False'

what to do if it is an old schema?

**Comment by mohanish2504** [comment]
Hi,
I want to contribute and I am a beginner, I wanted to know is the issue resolved? I am stuck here

`if is_old_schema(path): return False`

what to do if it is an old schema?

**Comment by Luis-gd** [comment]
Can i take this work?

**Comment by marco-c** [comment]
> I want to contribute and I am a beginner, I wanted to know is the issue resolved? I am stuck here
> if is_old_schema(path): return False

No, the issue is still here. If you look at the code, you'll see that `is_old_schema` is not checking the schema version of the downloaded DB, but the schema version of the remote DB.

> what to do if it is an old schema?

Please read all the comments, it is explained above.

**Comment by marco-c** [comment]
> Can i take this work?

You can work on any issue which is open and has no PR linked to it.
Did you read https://github.com/mozilla/bugbug/issues/1092 or CONTRIBUTING.md?

**Comment by avijit1258** [comment]
Hello @marco-c ,

I am thinking to implement a new function download_check_version() similar to download_check_etag() in the bugbug/utils.py which will be called after downloading the current DB. Then return false if that is the case. Am I going to the right direction?

Avijit

**Comment by marco-c** [comment]
@avijit1258 the goal is to remove a DB when its schema version is old compared to the current schema. I'm not sure how what you proposed is going to do this, but I'm not sure exactly what you mean. Proposing a PR is the best way to show what you mean ;)

**Comment by avijit1258** [comment]
@marco-c  How can I reproduce the situation you are telling? Can you give any url to try or way to figure out this?
thanks

**Comment by marco-c** [comment]
@avijit1258 the first step is to look at the code I linked at in the first comment, and take a look at all the functions it calls and figure out how it works.

**Comment by avijit1258** [comment]
@marco-c  As you suggested I have gone through download(), is_old_schema(), download_check_etag() and extract_file(). I need some clarification. After calling the download_check_etag() function, download() function calls  extract_file(). I am wondering where download() downloaded the DB. 

In the first comment of download function, it says 

> If a DB with the current schema is not available yet, we can't download.
Here current means new or latest? or existing? 

Another question is if I am getting the issue right the purpose is to ensure the DB schema is not updated while we are downloading the DB. For that, we are planning to check the DB schema after downloading with the latest schema version of the server?

Thanks

`def download(path, support_files_too=False):
    # If a DB with the current schema is not available yet, we can't download.
    print('Path: ', path)
    if is_old_schema(path):
        return False

    zst_path = f"{path}.zst"

    url = DATABASES[path]["url"]
    try:
        logger.info(f"Downloading {url} to {zst_path}")
        updated = utils.download_check_etag(url, zst_path)

        if updated:
            extract_file(zst_path)

        successful = True
        if support_files_too:
            for support_file in DATABASES[path]["support_files"]:
                successful |= download_support_file(path, support_file)

        return successful
    except requests.exceptions.HTTPError:
        logger.info(f"{url} is not yet available to download", exc_info=True)
        return False`

**Comment by avijit1258** [comment]
@marco-c I have added a PR with my proposed solution. #1206

**Comment by marco-c** [comment]
No, the purpose is to avoid using an already downloaded DB if its schema is not the current schema. The schema of the downloaded DB can be read from the ".version" file, its current schema can be read from the DATABASES dict.
Thinking about this more, I think this is basically already covered by the etag check. So we just need to write a test.

**Comment by avijit1258** [comment]
@marco-c Thanks for the clarification. Things are becoming more clear now. By saying test, do you mean checking it in the download function or writing a test case for download function in test file?

**Comment by marco-c** [comment]
> @marco-c Thanks for the clarification. Things are becoming more clear now. By saying test, do you mean checking it in the download function or writing a test case for download function in test file?

The second: writing a test case.

**Comment by abhaykatheria** [comment]
Can I just simply assert that new db is not downloaded 
assert updated '' new db not downloaded"
And then in test I can write case where the current dB is older but newer can't be downloaded.

**Comment by sakshamb2113** [comment]
@marco-c can i work on this issue?I am new to open source and want to contribute.

**Comment by marco-c** [comment]
Yes, please read CONTRIBUTING.md and https://github.com/mozilla/bugbug/issues/1092 for the contributing rules.

**Comment by tarunjarvis5** [comment]
But isn't line 91 already check if its old schema and therefore if it is then returning false
line88:# Download and extract databases.
line89:def download(path, support_files_too=False):
line90:    # If a DB with the current schema is not available yet, we can't download.
line91:    if is_old_schema(path):
line92:        return False

**Comment by marco-c** [comment]
@tarunjarvis5 yes, the code is doing that, but this issue is about adding a test to ensure it works correctly.

**Comment by BenDeBrasi** [comment]
Hi, can I work on this?

**Comment by marco-c** [comment]
Yes, please read CONTRIBUTING.md and #1092 for the contributing rules.

**Comment by timalsinab** [comment]
Hello, I am an outreachy applicant. Can I work on it?

**Comment by suhaibmujahid** [comment]
> Hello, I am an outreachy applicant. Can I work on it?

Yes, please read [CONTRIBUTING.md](https://github.com/mozilla/bugbug/blob/master/CONTRIBUTING.md) and https://github.com/mozilla/bugbug/issues/1092 for the contributing rules.

**Comment by fadebowaley** [comment]
I'm interested in this, please confirm if its fine to add a new file for this test. Thanks

**Comment by suhaibmujahid** [comment]
> I'm interested in this, please confirm if its fine to add a new file for this test. Thanks

As stated in [CONTRIBUTING.md](https://github.com/mozilla/bugbug/blob/master/CONTRIBUTING.md), you can work on any issue without asking, as long as there is no open PR linked to it.

**Comment by KhadijaKamran** [comment]
Hey @marco-c, Do I need to write assert statements for testing this function?

**Comment by suhaibmujahid** [comment]
> Hey @marco-c, Do I need to write assert statements for testing this function?

Yes, please read the comments above.

**Comment by fadebowaley** [comment]
Workng on Bug ID id 1108
 I checked  the [db.py](http://db.py/) and also i understand  the problem . How to write the test file ?. I noticed a tests folder called 'tests' is this where the solution will be? Will I use pytest to run this ?

I also want to discover how to run the test (I have already installed -test-requirements.txt

**Comment by fadebowaley** [comment]
@suhaibmujahid  To make sure a new DB is downloaded, I have assertion error while atTempting to download the database 

```
............
   assert  db.is_different_schema(db_path) 
    assert not db.download(db_path) # the assertion error comes here 

    assert not os.path.exists(db_path)
    assert not os.path.exists(db_path.with_suffix(db_path.suffix + ".zst"))
    assert not os.path.exists(db_path.with_suffix(db_path.suffix + ".zst.etag"))
```

```
............................./pytest-35/test_download_newer_schema0/prova.json
=========================================================== short test summary info ============================================================
FAILED tests/test_db.py::test_download_newer_schema - AssertionError: assert False
======================================================== 1 failed, 35 passed in 58.69s =========================================================
```

............

**Comment by fadebowaley** [comment]
> @fadebowaley are you still working on this?

As stated in [CONTRIBUTING.md](https://github.com/mozilla/bugbug/blob/master/CONTRIBUTING.md), you can work on any issue without asking, as long as there is no open PR linked to it.
